lib/Jsrt/Jsrt.cpp

```diff
@@ -2746,3 +3230,504 @@ CHAKRA_API JsRunSerializedScriptWithCallback(_In_ JsSerializedScriptLoadSourceCa
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+    ThreadContext* threadContext = scriptContext->GetThreadContext();
+    if(threadContext->TTDLog == nullptr)
+    {
+        AssertMsg(false, "Need to create in TTD mode.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDDetached())
+    {
+        AssertMsg(false, "Cannot re-start TTD after detach.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDActive())
+    {
+        AssertMsg(false, "Already started TTD.");
+        return JsErrorCategoryUsage;
+    }
+
+    threadContext->TTDLog->SetGlobalMode(TTD::TTDMode::RecordEnabled);
+
+    threadContext->TTDLog->PushMode(TTD::TTDMode::ExcludedExecution);
+    BEGIN_JS_RUNTIME_CALLROOT_EX(scriptContext, false)
+    {
+        threadContext->TTDLog->DoSnapshotExtract();
+    }
+    END_JS_RUNTIME_CALL(scriptContext);
+    threadContext->TTDLog->PopMode(TTD::TTDMode::ExcludedExecution);
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDStopTimeTravelRecording()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+    ThreadContext* threadContext = scriptContext->GetThreadContext();
+    if(threadContext->TTDLog == nullptr)
+    {
+        AssertMsg(false, "Need to create in TTD mode.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDDetached())
+    {
+        AssertMsg(false, "Already stopped TTD.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(!scriptContext->IsTTDActive())
+    {
+        AssertMsg(false, "TTD was never started.");
+        return JsErrorCategoryUsage;
+    }
+
+    threadContext->TTDLog->PushMode(TTD::TTDMode::ExcludedExecution);
+    BEGIN_JS_RUNTIME_CALLROOT_EX(scriptContext, false)
+    {
+        threadContext->EmitTTDLogIfNeeded();
+        threadContext->EndCtxTimeTravel(scriptContext);
+    }
+    END_JS_RUNTIME_CALL(scriptContext);
+    threadContext->TTDLog->PopMode(TTD::TTDMode::ExcludedExecution);
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDEmitTimeTravelRecording()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+    ThreadContext* threadContext = scriptContext->GetThreadContext();
+    if(threadContext->TTDLog == nullptr)
+    {
+        AssertMsg(false, "Need to create in TTD mode.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDDetached())
+    {
+        AssertMsg(false, "Already stopped TTD.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(!scriptContext->IsTTDActive())
+    {
+        AssertMsg(false, "TTD was never started.");
+        return JsErrorCategoryUsage;
+    }
```

Comment:
Same checks/code in multiple APIs, can we put this in a macro similar to VALIDATE_RUNTIME_STATE_FOR_START_STOP_DEBUGGING